File: 745-find-smallest-letter-greater-than-target/find-smallest-letter-greater-than-target.java
class Solution {
    // letters = ["c","f","j"], target = "a"
    // letters = ["c","f","j"], target = "c"
    // letters = ["x","x","y","y"], target = "z"
    // time: O(n) 
    // space: O(1)
    public char nextGreatestLetter(char[] letters, char target) {
        for (char c : letters) {
            if (Character.compare(c, target) > 0) {
                return c;
            }
        }
        return letters[0];
    }
}

File: 745-find-smallest-letter-greater-than-target/find-smallest-letter-greater-than-target.py
class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        # for letter in letters:
        #     if letter > target:
        #         return letter
        
        # return letters[0]
        
        # time: O(n)
        # space: O(1)

        left, right = 0, len(letters) - 1

        while left <= right:
            middle = (left + right) // 2

            if letters[middle] <= target:
                left = middle + 1
            else:
                right = middle - 1
        
        if left == len(letters):
            return letters[0]
        else:
            return letters[left]
        